drivers/accel/adxl355/adxl355.c

```diff
@@ -0,0 +1,814 @@
+int adxl355_read_device_data(struct adxl355_dev *dev, uint8_t base_address,
+			     uint16_t size, uint8_t *read_data)
+{
+	int status = 0;
+
+	if (dev->communication_type == ADXL355_SPI_COMM) {
+		uint8_t *comm_buffer;
+		comm_buffer = (uint8_t *)calloc(1,1 + size);
+
+		comm_buffer[0] = ADXL355_SPI_READ | (base_address << 1);
+		status = spi_write_and_read(dev->spi_desc, comm_buffer, 1 + size);
+		for (uint16_t idx = 0; idx < size; idx++)
+			read_data[idx] = comm_buffer[idx+1];
+
+		free(comm_buffer);
+	} else {
+		status = i2c_write(dev->i2c_desc, &base_address, 1, 0);
+		if (status != 0)
+			return -1;
+		status = i2c_read(dev->i2c_desc, read_data, size, 1);
+	}
+
+	return status;
+}
+
+/***************************************************************************//**
+ * @brief Writes to the device
+ *
+ * @param dev              - The device structure.
+ * @param base_address     - Address of the base register.
+ * @param size             - The number of bytes to be written. It is the size
+ *                           of the write_data buffer.
+ * @param write_data       - The data which is going to be written.
+ *
+ * @return status          - Result of the reading procedure.
+*******************************************************************************/
+int adxl355_write_device_data(struct adxl355_dev *dev, uint8_t base_address,
+			      uint16_t size, uint8_t *write_data)
+{
+	int status = 0;
+	uint8_t *comm_buffer;
+	comm_buffer = (uint8_t *)calloc(1,1 + size);
+	for (uint16_t idx = 0; idx < size; idx++)
+		comm_buffer[1+idx] = write_data[idx];
+
+	if (dev->communication_type == ADXL355_SPI_COMM) {
+		comm_buffer[0] = ADXL355_SPI_WRITE | (base_address << 1);
+		status = spi_write_and_read(dev->spi_desc, comm_buffer, size + 1);
+	} else {
+		comm_buffer[0] = base_address;
+		status = i2c_write(dev->i2c_desc, comm_buffer, size + 1, 1);
+	}
+
+	free(comm_buffer);
+
+	return status;
+}
+
+/***************************************************************************//**
+ * @brief Initializes the communication peripheral and checks if the ADXL355
+ *        part is present.
+ *
+ * @param device     - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       		   parameters.
+ *
+ * @return status    - Result of the initialization procedure.
+ *                     Example: -1 - I2C/SPI peripheral was not initialized or
+ *                                   ADXL355 part is not present.
+ *                               0 - I2C/SPI peripheral is initialized and
+ *                                   ADXL355 part is present.
+*******************************************************************************/
+int adxl355_init(struct adxl355_dev **device,
+		 struct adxl355_init_param init_param)
+{
+	struct adxl355_dev *dev;
+	int status = 0;
+	uint8_t reg_value = 0;
+
+	dev = (struct adxl355_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
```

Comment:
We're either using 0 or errno(3) error codes as function return values (this way the user can get more info about what went wrong). In this case you could return -ENOMEM (notice the -).